kv: add per-batch limit on scan results

RaduBerinde opened the pull request:
To scan tables in batches (#2966) when there are multiple spans, we need a
per-batch limit on the total number of scan results for all scans and reverse
scans in the batch. This change adds this limit as a parameter in the
BatchRequest header.

Addresses #4696.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5092)

<!-- Reviewable:end -->

tschottdorf commented:
Adding this logic in `Replica` is unfortunately not - it needs to live in `DistSender`, which is where the batch gets sent to the various involved ranges. Querying range after range, you'd check the max limit and update the individual scan limits accordingly. Current behavior should be pretty broken if you actually have multiple ranges touched by a query, so there need to be more tests (there are `TestMultiRangeScan` tests or something like that somewhere).

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[client/batch.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWlddklocEyXBGsWOl:-KCWlddlcpCmEfw3nkVx:-1683393328) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/client/batch.go#L49)):*
Would this more naturally apply to all Limit'able operations within a batch?

---

*[client/client_test.go, line 410 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWm3XcA67N-d1JMjwc:-KCWm3XcA67N-d1JMjwd:-2145331817) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/client/client_test.go#L410)):*
If you don't mind the archeology, could you see about this guy? https://github.com/cockroachdb/cockroach/issues/1891 is closed.

---

*[client/client_test.go, line 444 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWmHuutEf4esnVbSoe:-KCWmHuutEf4esnVbSof:1119110271) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/client/client_test.go#L444)):*
that wording suggests two individual spans with limit. Better:

> Try a limited batch consisting of two spans.

---

*[client/client_test.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWmVAZAnvcypKtYXs9:-KCWmVAZAnvcypKtYXsA:-1589557173) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/client/client_test.go#L447)):*
Try one case where you batch them up in opposite order.

---

*[client/client_test.go, line 480 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWmarX1YrD0UCPXJ8Y:-KCWmarX1YrD0UCPXJ8Z:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/client/client_test.go#L480)):*
ditto.

---

*[storage/replica.go, line 1581 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWoUio6OCqLBwtoNLM:-KCWoUio6OCqLBwtoNLN:-841329537) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/storage/replica.go#L1581)):*
this is an allocation which is always in the critical path. Use a pointer and `nil` to communicate "no limit".

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCWmtOEGS_MenxDKSSo:-820431996)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ah, I understand. Obviously I am a noob wrt this code :) Will move the logic in DistSender

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCWsYqmQ1PuBXYJLsw6:1108570158)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
No worries. Moving isn't enough btw, you'll still need what you have in
Replica (unless you're willing to live with sometimes truncating obtained
results in DistSender in cases where more than one Span are active on a
Range and something which is above-limits comes back).

On Thu, Mar 10, 2016 at 1:59 PM RaduBerinde <notifications@github.com>
wrote:

> Ah, I understand. Obviously I am a noob wrt this code :) Will move the
> logic in DistSender
> ------------------------------
>
> Review status: all files reviewed at latest revision, 6 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCWsYqmQ1PuBXYJLsw6:1108570158>*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5092#issuecomment-195001459>
> .
>

RaduBerinde commented:
Right, makes sense. I see the existing code in `sendChunk` that does this when one scan spans ranges.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1581 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWoUio6OCqLBwtoNLM:-KCWvUFEQLXl9P7ehPB4:-31839429) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/storage/replica.go#L1581)):*
I noticed the code in DistSender that uses `Countable`. Do you think it would be cleaner to do the same here? Instead of passing a pointer to executeCmd we could check for a `Countable` result and update it ourselves.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCWuK3gM-cRWu1VbIjF:1082863167)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1581 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCWoUio6OCqLBwtoNLM:-KCWxdDGgLH2D5wIMpDh:34136295) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d1600c4286780249de2687510b3cba606f25e2b/storage/replica.go#L1581)):*
yes, I think that makes sense (that would mean that this max limit is applied unconditionally to all limitable things in the batch -  we'll definitely have to sit down and get proper semantics in place at some point - for example, after this change you can't just pass a batch to someone else unless you also tell them you put a limit on).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Sounds good, just make sure to sprinkle comments liberally. One day this'll grow into something robust.

RaduBerinde commented:
PTAL

---

Review status: 6 of 15 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCbMswFA-j9_a8bvQWt:676965300)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Got to run mid-review, but left some comments.

---

Review status: 6 of 15 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[client/batch.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbdD1oMMDe52UeKAX8:-KCbdD1prm3PBz7epVH4:1230300792) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/batch.go#L49)):*
s/ar/are/

---

*[client/util.go, line 125 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbeGKhwiV6g_WEJZ6E:-KCbeGKivsxpsUoI4SA9:-2044781614) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L125)):*
this looks like it should be `filepath.Basename()` or something like that.

---

*[client/util.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbe0waYOFhCdzhWnCI:-KCbe0wbOaas1Y9g7nDo:1076405237) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L133)):*
as far as I can tell this is never used from non-testing code. Why not move this back into client_test.go or rename to util_test?

---

*[kv/dist_sender.go, line 506 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbeggF5VVOAL6F2phX:-KCbeggGe9KHiEV5TMzA:607294362) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L506)):*
s/same-direction scans/MaxScanResults/

---

*[kv/dist_sender.go, line 565 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbem9C2iryJpmxM_Ci:-KCbem9C2iryJpmxM_Cj:350227926) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L565)):*
nit: `var batchLimit bool`

---

*[kv/dist_sender.go, line 566 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbepF09TN1yXOHG0Gb:-KCbepF09TN1yXOHG0Gc:240666773) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L566)):*
nit: `var remScanResults int64`

---

*[kv/dist_sender.go, line 569 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbgTeXn_1WL-FDwT39:-KCbgTeYJP9nIjLj823H:1711096574) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L569)):*
`ba` is a copy, so you can freely modify `MaxScanResults`. There are just a lot of variables here; it seems that you should be able to use `ba.MaxScanResults` at all times (no `remScanResults`, no `batchLimit`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCbhYqldcwjbrDqj919:-1987764586)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 6 of 15 files reviewed at latest revision, 12 unresolved discussions.

---

*[client/util.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbe0waYOFhCdzhWnCI:-KCcUQuqVvasEF7HGA2p:-2073065194) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L133)):*
It is used from a test in another package (`kv/dist_sender_server_test.go`). I couldn't find a way to import the `client_test` package from there. Moving to util_test doesn't work either. If you know of a way to do one of these or have other ideas let me know.

---

*[kv/dist_sender.go, line 506 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbeggF5VVOAL6F2phX:-KCcRd-6_MLMDvVQLlpu:844305285) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L506)):*
This condition cannot happen because the batch contains only scans (which we verified above), which should never need splitting. Added a comment to clarify.

---

*[kv/dist_sender.go, line 569 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbgTeXn_1WL-FDwT39:-KCcSWuCdxkvuViyG6kl:1292868569) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L569)):*
The issue is that `remScanResults=0` means we have no more results left to return, whereas `MaxScanResults=0` means no limit. I don't want to mix and match these different semantics in a single variable.

I can get rid of `batchLimit` and init `remScanResults` to `MaxInt64` if you prefer that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 9 of 9 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[client/util.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbe0waYOFhCdzhWnCI:-KCcdGRm2k3hFgSuett_:-257227859) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L133)):*
Ah, ok. No, not possible then. Maybe `testutils` is a good new home?

---

*[kv/dist_sender.go, line 506 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbeggF5VVOAL6F2phX:-KCcdWrztrvAkzS6rh4T:87402724) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L506)):*
I know it can't happen, just clarified the message.

---

*[kv/dist_sender.go, line 569 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbgTeXn_1WL-FDwT39:-KCcdo45k763HISw92-U:738342827) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L569)):*
But you return immediately when you hit zero, right? So your code below after `Combine()` could be

```go

if ba.MaxScanResults > 0 {  // have an active limit
  // count
  remResults := ba.MaxScanResults - numResults
  if remResults <= 0 {
    // do the no-op stuff and return
  }
  ba.MaxScanResults = remResults // still > 0
}
```

I think that would be clearer (since no new vars creep up outside), but maybe this is being to clever, in which case just do what you think is best.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 14 of 15 files reviewed at latest revision, 11 unresolved discussions.

---

*[client/util.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbe0waYOFhCdzhWnCI:-KCcfRgvBnpPGuY8Fn9P:-545102737) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L133)):*
That's where I started but it needs to use `client.KeyValue` and I can't import `client` from `testutils`

---

*[kv/dist_sender.go, line 569 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbgTeXn_1WL-FDwT39:-KCcgI8DKzn4yQbuAz_r:-716185254) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/kv/dist_sender.go#L569)):*
Ah, duh, you're right, done. Much better this way.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 9 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/util.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-KCbe0waYOFhCdzhWnCI:-KCd4qrc7DsvKb1LISDb:-1610723275) ([raw file](https://github.com/cockroachdb/cockroach/blob/738399272f96b1918b8f357a357d965e9941e4d8/client/util.go#L133)):*
Ah, ok. Well, looks like this is where it's gonna be.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5092#-:-KCd58kZherHNgXA_qxK:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
kv: add per-batch limit on scan results

To scan tables in batches (#2966) when there are multiple spans, we need a
per-batch limit on the total number of scan results for all scans and reverse
scans in the batch. This change adds this limit as a parameter in the
BatchRequest header.

Addresses #4696.
